File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
#define i long long
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        i ans = 0;

        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});

        while (!q.empty()) {
            int n = q.size();
            int minIndex = q.front().second;
            i  first,last;
            for (int j = 0; j < n; j++) {
                auto it = q.front();
                int curr_index = it.second - minIndex;
                q.pop();
                if(j == 0){
                    first = curr_index;
                }
                if(j==n-1){
                    last = curr_index;
                }

                if(it.first->left){
                    q.push({it.first->left,2*curr_index+1});
                }
                if(it.first->right){
                    q.push({it.first->right,2*curr_index+2});
                }
            }
            ans = max(ans,last-first+1);
        }
        return ans;
    }
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func widthOfBinaryTree(root *TreeNode) int {
	type BT struct {
		Node  *TreeNode
		Index int64
	}
	ans := int64(0)
	q := []BT{{root, 0}}

	for len(q) > 0 {
		minIndex := q[0].Index
		n := len(q)
		var first, last int64
		for i := 0; i < n; i++ {
			currNode := q[0]
			q = q[1:]
			currIndex := currNode.Index - minIndex

			if i == 0 {
				first = currIndex
			}
			if i == n-1 {
				last = currIndex
			}

			if currNode.Node.Left != nil {
				q = append(q, BT{currNode.Node.Left, 2*currIndex + 1})
			}
			if currNode.Node.Right != nil {
				q = append(q, BT{currNode.Node.Right, 2*currIndex + 2})
			}
		}
		ans = max(ans, last-first+1)
	}
	return int(ans)
}
func max(a, b int64) int64 {
	if a >= b {
		return a
	}
	return b
}